### PR TITLE
Refactor /decide enpoint & allow recording without autocapture

### PR DESCRIPTION
**Pull request by paolodamico**
## Changes

This PR enables running session recording even if autocapture is disabled. This is motivated by a couple of requests from different users who wish to use session recording without autocapture.
- To enable this, we refactor the `/decide` endpoint logic out of the `autocapture` module.

## Checklist
- [X] Tests for new code (if applicable)
- [X] TypeScript definitions (module.d.ts) updated and in sync with library exports. **Not applicable**

### CODE REVIEW

File: cypress/integration/session-recording.spec.js [hunk 1]
```diff
@@ -8,7 +8,7 @@ describe('Session recording', () => {
             method: 'POST',
             url: '**/decide/*',
             response: {
-                config: { enable_collect_everything: true },
+                config: { enable_collect_everything: false },
```
Comment (macobo) [inline cypress/integration/session-recording.spec.js:11]: Q: Why this change?

File: src/decide.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import { autocapture } from './autocapture'
+import { _ } from './utils'
+
+const decide = {
```
Comment (macobo) [inline src/decide.js:4]: Code style nit: Let's use classes as with `src/extensions/sessionrecording.js`.

So constrctor + `callDecide` function.

File: src/posthog-core.js [hunk 3]
```diff
@@ -131,6 +132,8 @@ var create_mplib = function (token, config, name) {
     instance.sessionRecording = new SessionRecording(instance)
     instance.sessionRecording.startRecordingIfEnabled()
 
+    decide.init(instance)
```
Comment (macobo) [inline src/posthog-core.js:135]: So this refactoring has a non-intuitive side-effect which might be considered a regression - we now always call /decide. 

Previously decide was not called when autocapture was off - making the script a lot more light-weight if nothing was turned on.

File: src/decide.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            token: instance.get_config('token'),
+            distinct_id: instance.get_distinct_id(),
+        })
+
+        var encoded_data = _.base64Encode(json_data)
+        instance._send_request(
+            instance.get_config('api_host') + '/decide/',
+            { data: encoded_data },
+            { method: 'POST' },
+            instance._prepare_callback(this.parseDecideResponse)
```
Comment (macobo) [inline src/decide.js:19]: Note: `_prepare_callback` is not really needed. posthog-js does not really work unless you use xhr.

File: src/decide.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    instance: null,
+
+    init: function (instance) {
+        this.instance = instance
+        var json_data = JSON.stringify({
+            token: instance.get_config('token'),
+            distinct_id: instance.get_distinct_id(),
+        })
+
+        var encoded_data = _.base64Encode(json_data)
```
Comment (macobo) [inline src/decide.js:14]: es6 please (var => const)!

File: src/decide.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        instance._send_request(
+            instance.get_config('api_host') + '/decide/',
+            { data: encoded_data },
+            { method: 'POST' },
+            instance._prepare_callback(this.parseDecideResponse)
+        )
+    },
+
+    parseDecideResponse(response) {
```
Comment (macobo) [inline src/decide.js:23]: This is missing tests - should we add them?

**Review by macobo — COMMENTED** [review]
Note - the idea seems good, one worry about a regression below.

However from a product perspective is this a case of perverse incentives since we do not bill session recordings? Load and cost-wise session recordings >>>> autocapture for us.

File: src/posthog-core.js (hunk 3 above)
Comment (paolodamico) [reply src/posthog-core.js:135]: Ack you're right. Well definitely something worth changing would be to not call the `/decide` endpoint if session recording and autocapture are both disabled. Perhaps if we want to avoid the extra request to `/decide` we need to introduce a different param that enables session recording even if autocapture is off? Thoughts?

File: src/decide.js (hunk 2 above)
Comment (paolodamico) [reply src/decide.js:4]: Ack had it as a class first but thought it was more straightforward like this, reverting to class 😉

File: src/posthog-core.js (hunk 3 above)
Comment (paolodamico) [reply src/posthog-core.js:135]: Per sync conversation, we'll keep the new approach as `/decide` is also required for other stuff like compression and feature flags, and adding an advanced config option to fully disable this request.

File: src/decide.js (hunk 6 above)
Comment (paolodamico) [reply src/decide.js:27]: Done!

File: cypress/integration/session-recording.spec.js (hunk 1 above)
Comment (paolodamico) [reply cypress/integration/session-recording.spec.js:11]: Intended to make sure that session recording works even if autocapture is off

**Comment by paolodamico** [comment]
I've updated the PR based on what we talked about and is ready for a review. Will create a complimentary PR to update the relevant docs. Not sure why the testcafe tests are failing, can you give me some pointers on how to debug? Can't seem to find any particularly useful information on Browserstack.

File: src/__tests__/compression.js [hunk 7]
```diff
@@ -98,6 +99,8 @@ describe('Payload Compression', () => {
         })
 
         it('should save supported compression in instance', () => {
+            const decide = new Decide(lib)
```
Comment (macobo) [inline src/__tests__/compression.js:102]: Note: you can shorten this to `new Decide(lib).call()`

File: src/__tests__/decide.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            afterDecideResponse: jest.fn(),
+        },
+        persistence: { register: jest.fn(), unregister: jest.fn() },
+    }))
+
+    given('decideResponse', () => ({ enable_collect_everything: true }))
+
+    given('config', () => ({ api_host: 'https://test.com' }))
+
+    jest.spyOn(autocapture, 'afterDecideResponse').mockImplementation()
```
Comment (macobo) [inline src/__tests__/decide.js:32]: This should be in a beforeEach block.

File: src/posthog-core.js [hunk 9]
```diff
@@ -722,7 +732,13 @@ PostHogLib.prototype._register_single = function (prop, value) {
  * @param {Object|String} options (optional) If {send_event: false}, we won't send an $feature_flag_call event to PostHog.
  */
 PostHogLib.prototype.isFeatureEnabled = function (key, options = {}) {
-    return this.feature_flags.isFeatureEnabled(key, options)
+    if (!this.featureFlags) {
```
Comment (macobo) [inline src/posthog-core.js:735]: This change has a couple of issues:

1. Returning undefined makes no sense and can break peoples code. This function should return a boolean. `false` is the good default
2. Checking `!this.featureFlags` is misleading as it will fail if no flags are turned on.
3. This will cause logspam - this is always reached if isFeatureEnabled is called before decide returns a response.
4. This doesn't belong to posthog-core.js at all - feature flags file should handle all the logic here. Otherwise the two ways to call featureflags in the library diverge in the implementation.

I suggest not including this change at all.

File: src/posthog-core.js [hunk 10]
```diff
@@ -135,6 +137,24 @@ var create_mplib = function (token, config, name) {
     instance['people'] = new PostHogPeople()
     instance['people']._init(instance)
 
+    // if any instance on the page has debug = true, we set the
+    // global debug to be true
+    Config.DEBUG = Config.DEBUG || instance.get_config('debug')
+
+    // if target is not defined, we called init after the lib already
+    // loaded, so there won't be an array of things to execute
+    if (!_.isUndefined(target) && _.isArray(target)) {
```
Comment (macobo) [inline src/posthog-core.js:146]: Moving this code is very likely to break something - it has to do with support for if posthog-js is included multiple times and there are many different entrypoints to the script. I would err on being paranoid here - does this _need_ to be moved?

File: src/posthog-core.js [hunk 11]
```diff
@@ -135,6 +137,24 @@ var create_mplib = function (token, config, name) {
...
+        // Crunch through the people queue first - we queue this data up &
+        // flush on identify, so it's better to do all these operations first
+        instance._execute_array.call(instance['people'], target['people'])
+        instance._execute_array(target)
+    }
+
+    if (instance.get_config('advanced_disable_decide')) {
+        // When /decide endpoint is disabled, nonone of the features below are available.
+        return instance
```
Comment (macobo) [inline src/posthog-core.js:155]: This is a breaking change. We advertise `posthog.featureFlags` in our documentation, this will make that code throw if a flag is on. 

Suggestion: just gate the decide call.

File: src/posthog-core.js [hunk 12]
```diff
@@ -144,9 +164,8 @@ var create_mplib = function (token, config, name) {
     instance.sessionRecording = new SessionRecording(instance)
     instance.sessionRecording.startRecordingIfEnabled()
 
-    // if any instance on the page has debug = true, we set the
-    // global debug to be true
-    Config.DEBUG = Config.DEBUG || instance.get_config('debug')
+    instance.decide = new Decide(instance)
```
Comment (macobo) [inline src/posthog-core.js:167]: Why do we save this on the instance? I don't think there's a real reason to.

**Review by macobo — COMMENTED** [review]
Added some notes. Tests seem to be failing as well.

File: src/posthog-core.js (hunk 9 above)
Comment (paolodamico) [reply src/posthog-core.js:735]: Removed this code, users will get an error if attempting to load feature flags with `/decide` endpoint disabled anyways.

File: src/posthog-core.js (hunk 11 above)
Comment (paolodamico) [reply src/posthog-core.js:155]: Agreed, per discussion on Slack

**Comment by paolodamico** [comment]
I've addressed all your feedback @macobo, thanks for the thorough review!

I've spent quite a bit of time trying to track down why the testcafe tests are not passing but haven't gotten very far. What's worse is that sometimes the Chrome tests pass, sometimes some of them, sometimes they don't. The only suspect I have is that the /decide call is now being treated async whereas before some condition in the way we run tests was causing it to run synchronously. Can we have a quick sync session to try to figure this out together @macobo ?

**Comment by macobo** [comment]
There are some issues with ingestion not working as it did when the e2e test was added - not on this PR. Will be investigating and fixing as soon as cycles free up!

**Comment by paolodamico** [comment]
+1 from a user wanting to use the toolbar while having autocapture off
